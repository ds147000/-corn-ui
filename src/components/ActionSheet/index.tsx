/* eslint-disable import/export */

import ActionSheet  from './actionSheet'

export { ActionSheetItem } from './item'
export default ActionSheet
