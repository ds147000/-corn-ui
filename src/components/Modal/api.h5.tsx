/* eslint-disable no-magic-numbers */
import  React, { useState } from 'react'
import Modal, { ModalButton } from './modal'
import Protal from '../Portal'

export interface ModalOption {
  /** 提示的标题 */
  title?: string
  /** 取消按钮的文字，最多 4 个字符 */
  cancelText?: string
  /** 确认按钮的文字，最多 4 个字符 */
  confirmText?: string
  /** 是否显示取消按钮 */
  showCancel?: boolean
  /** 提示的内容 */
  content?: string
}

export interface SuccessCallbackResult {
  /** 为 true 时，表示用户点击了取消（用于 Android 系统区分点击蒙层关闭还是点击取消按钮关闭） */
  cancel: boolean
  /** 为 true 时，表示用户点击了确定按钮 */
  confirm: boolean
  /** 调用结果 */
  errMsg: string
}

interface DomProps extends ModalOption {
  onClick(SuccessCallbackResult): void
  onHide(): void
}

const Dom: React.FC<DomProps> = ({
  title,
  cancelText = '取消',
  confirmText = '确定',
  showCancel,
  content,
  onClick,
  onHide
}) => {
  const [ show, setShow ] = useState(true)
  const button: ModalButton[] = [
    { text: confirmText, style: 'success' },
  ]

  if (showCancel) button.push({ text: cancelText, style: 'cancel' })

  const onButtonClick = (index: number): void => {
    setShow(false)
    onClick?.({
      confirm: index === 0,
      cancel: index === 1,
      errMsg: 'ok'
    })
  }


  return (
    <Modal
      title={title}
      content={content}
      visible={show}
      onButtonClick={onButtonClick}
      button={button}
      onHide={onHide}
    />
  )
}

const showModal = async (option: ModalOption): Promise<SuccessCallbackResult> => {
  // eslint-disable-next-line no-async-promise-executor
  return new Promise(async (res, rej) => {
    let id: number

    const remove = (): void => {
      Protal.remove(id)
    }

    try {
      id = await Protal.add(<Dom {...option} onHide={remove} onClick={res} />)
    } catch (error) {
      rej({ errMsg: error.message } as SuccessCallbackResult)
    }
  })
}

export default showModal
