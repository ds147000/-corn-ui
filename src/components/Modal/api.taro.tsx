import { showModal } from '@tarojs/taro'

const _showModal = async (option: showModal.Option): Promise<showModal.SuccessCallbackResult> => {
  option.cancelColor = '#818894'
  option.confirmColor = '#00A9FF'
  return showModal(option)
}

export default _showModal
