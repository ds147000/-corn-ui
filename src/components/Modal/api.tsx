/* eslint-disable import/export */
// #if _APP === 'weapp'
import showModal from './api.taro'

export default showModal

// #else
// eslint-disable-next-line import/first
import showActionSheet from './api.h5'

export default showActionSheet
// #endif
