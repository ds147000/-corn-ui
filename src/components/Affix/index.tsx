/* eslint-disable no-unreachable */
/* eslint-disable no-magic-numbers */
// #if _APP === 'weapp'
import Taro, { usePageScroll } from '@tarojs/taro'
// #endif
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { View } from '@tarojs/components'
import ClassNames from 'classnames'
import { ViewProps } from '../../types/View'
import { judge } from './utils/judge'

export interface AffixProps extends ViewProps {
  position?: 'top' | 'bottom' | 'right' | 'left'
  onChange?(fixed: boolean): void
}

const Affix: React.FC<AffixProps> = ({ children, position = 'top', onChange, className, ...props }) => {
  const [ fixed, setFixed ] = useState(false)
  const [ boxStyle, setBoxStyle ] = useState<React.CSSProperties>({})
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const ref = useRef<any>()

  const oldFixed = useRef<boolean>(false)

  const _class = useMemo(() => {
    return ClassNames('corn-affix', fixed && `corn-affix-${position}`, className)
  }, [ fixed, position, className ])

  const changeFixed = useCallback((value: boolean): void => {
    setFixed((state) => {
      oldFixed.current = state
      return value
    })
  }, [])

  // #if _APP === 'weapp'
  const system = useRef(Taro.getSystemInfoSync())
  const timer = useRef<NodeJS.Timeout>()

  const checkFlexd = (): void => {
    const node = ref.current

    const rootRect = {
      top: 0,
      left: 0,
      bottom: system.current.windowHeight,
      right: system.current.windowWidth
    } as DOMRectReadOnly

    Taro.createSelectorQuery()
      .select('#' + node.uid)
      .fields({ size: true, rect: true }, (res) => {
        if (!res) return

        setBoxStyle({ height: res?.height, width: res?.width })
        const result = judge(res as unknown as DOMRect, rootRect, position, 0)
        changeFixed(result)
      })
      .exec()
  }

  usePageScroll(() => {
    if (timer.current) clearTimeout(timer.current)
    timer.current = setTimeout(() => checkFlexd(), 10)
  })

  useEffect(() => {
    return (): void => {
      if (timer.current) clearTimeout(timer.current)
    }
  }, [])

  // #else
  useEffect(() => {
    // eslint-disable-next-line no-unreachable
    const el: HTMLDivElement = ref.current
    const rootBounds = {
      x: 0,
      y: 0,
      width: window.innerWidth,
      height: window.innerHeight,
      top: 0,
      left: 0,
      bottom: window.innerHeight,
      right: window.innerWidth
    } as DOMRectReadOnly

    setBoxStyle({ height: el.offsetHeight, width: el.offsetWidth })

    function handleScroll(): void {
      const item = el.getBoundingClientRect()
      const result = judge(item, rootBounds, position, 0)
      changeFixed(result)
    }
    handleScroll()
    window.addEventListener('scroll', handleScroll)
    return (): void => {
      window.removeEventListener('scroll', handleScroll)
    }
  }, [ position, changeFixed ])
  // #endif

  useEffect(() => {
    if (oldFixed.current !== fixed) {
      oldFixed.current = fixed
      onChange?.(fixed)
    }
  }, [ fixed, onChange ])

  return (
    <View style={boxStyle} ref={ref}>
      <View className={_class} {...props} >
        {children}
      </View>
    </View>
  )
}

export default Affix
