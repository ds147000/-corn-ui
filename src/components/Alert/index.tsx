import React from 'react'
import { View } from '@tarojs/components'
import classNames from 'classnames'
import { ViewProps } from '../../types/View'
import Icon from '../Icon'

const Alter: React.FC<ViewProps> = ({ children, className, ...props }) => {
  const _class = classNames(className, 'corn-alert', 'corn-f')
  return (
    <View className={_class} {...props} >
      <Icon name="error" className="icon" />
      {children}
    </View>
  )
}

export default Alter
