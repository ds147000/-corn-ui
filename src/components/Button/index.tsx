import React, { useMemo } from 'react'
import ClassNames from 'classnames'
import { Button as TaroButton } from '@tarojs/components'
import Link, { LinkProps } from '../Link'
import { ITouchEvent } from '../../types'

/** 按钮大小，max: 最大， big: 超大，large：大，middle：中，small：小，mini */
export type ButtonSize = 'max' | 'big' | 'large' | 'middle' | 'small' | 'mini'
export type ButtonType = 'primary' | 'warn' | 'error' | 'link' | 'pop' | 'default' | 'light' | 'stop'

export interface ButtonProps extends LinkProps {
  /** 按钮大小，max: 最大， big: 超大，large：大，middle：中，small：小，mini */
  size?: ButtonSize
  /** 按钮风格类型 */
  type?: ButtonType
  /** 图标 */
  icon?: React.ReactNode
  /** 将按钮宽度调整为其父宽度的选项	 */
  block?: boolean
  /** 镂空风格 */
  ghost?: boolean
  /** 是否禁用 */
  disabled?: boolean
  /** 跳转链接，填写启用A链接行为 */
  to?: string
  /** 表单点击行为 */
  formType?: 'submit' | 'reset'
  /** 开启固定宽度 */
  auto?: boolean
  /** 点击事件 */
  onClick?(e: ITouchEvent): void
}

const Button: React.FC<ButtonProps> = ({
  size = 'middle',
  type = 'primary',
  block,
  ghost,
  disabled,
  to,
  onClick,
  icon,
  children,
  className,
  formType,
  auto,
  ...props
}) => {
  const isIcon = icon !== undefined
  const isChildren = children !== undefined

  const _class = useMemo(() => {
    return ClassNames(
      'corn-btn',
      'corn-if corn-ac corn-jc',
      `corn-btn-${size}`,
      `corn-btn-${type}`,
      {
        'corn-btn-block': block,
        'corn-btn-ghost': ghost,
        'corn-btn-icon': isIcon && !isChildren,
        'corn-btn-child-icon': isIcon && isChildren,
        'corn-btn-disabled': disabled,
        'corn-btn-auto': auto
      },
      className
    )
  }, [ size, type, block, ghost, disabled, className, auto, isIcon, isChildren ])

  const _onClick = (e: ITouchEvent): void => {
    if (disabled) return

    onClick?.(e)
  }



  if (to) {
    return (
      <Link {...props} target="Text" to={to} className={_class} onClick={_onClick} disabled={disabled} type="default" >
        {children}
        {icon}
      </Link>
    )
  }

  return (
    <TaroButton {...props} formType={formType} className={_class} onClick={_onClick} >
      {children}
      {icon}
    </TaroButton>
  )
}

export default Button
