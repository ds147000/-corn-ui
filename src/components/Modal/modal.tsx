import React from 'react'
import { View } from '@tarojs/components'
import Drawer, { DrawerProps } from '../Drawer'

const MULTI_NUM = 3

export type ModalButton = {
  /** 按钮文字 */
  text: string | React.ReactNode
  /** 按钮风格, 默认 success */
  style?: 'cancel' | 'success'
  /** 按钮点击事件 */
  onClick?(): void
}

export interface ModalProps extends DrawerProps {
  /** 提示的标题 */
  title?: string | React.ReactNode
  /** 提示的内容 */
  content?: string | React.ReactNode
  /** 弹出按钮列表 */
  button: Array<ModalButton>
  /** 点击按钮事件 */
  onButtonClick?(index: number): void
}

const Modal: React.FC<ModalProps> = ({
  children,
  content,
  title,
  button,
  onButtonClick,
  ...props
}) => {
  const _centent = children || content
  const floorClass = button.length >= MULTI_NUM ? 'xrk-modal-floor-multi' : 'xrk-modal-floor-line'

  return (
    <Drawer
      {...props}
      position="center"
    >
      <View className="xrk-modal" data-testid="modal" >
        <View className="xrk-modal-top">
          <View className="xrk-modal-title">
            {title}
          </View>
          {Boolean(_centent) && (
            <View className="xrk-modal-content">
              {_centent}
            </View>
          )}

        </View>
        {/* 按钮 */}
        <View className={floorClass}>
          {button.map((item, key) => (
            <View
              className={`xrk-modal-button ${item.style}`}
              key={String(key)}
              onClick={(): void => {
                onButtonClick?.(key)
                item.onClick?.()
              }}
            >
              {item.text}
            </View>
          ))}
        </View>
      </View>
    </Drawer>
  )
}

export default Modal
