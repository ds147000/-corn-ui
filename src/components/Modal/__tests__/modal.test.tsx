/* eslint-disable no-magic-numbers */
import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import { View } from '@tarojs/components'
import Modal from '../modal'

describe('Modal', () => {

  test('basis', async () => {
    const screen = render(
      <Modal
        visible
        title="弹窗"
        button={[
          { text: '取消', style: 'cancel' },
          { text: '确定' }
        ]}
      />
    )
    await waitFor(() => screen.getByText('弹窗'))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('content', async () => {
    const screen = render(
      <Modal
        visible
        title="弹窗"
        content="我是弹窗"
        button={[
          { text: '取消', style: 'cancel' },
          { text: '确定' }
        ]}
      />
    )
    await waitFor(() => screen.getByText('弹窗'))
    await waitFor(() => screen.getByText('我是弹窗'))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })


  test('children', async () => {
    const screen = render(
      <Modal
        visible
        title="弹窗"
        button={[
          { text: '取消', style: 'cancel' },
          { text: '确定' }
        ]}
      >
        <View>我是弹窗的爸爸</View>
      </Modal>
    )
    await waitFor(() => screen.getByText('弹窗'))
    await waitFor(() => screen.getByText('我是弹窗的爸爸'))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('button', async () => {
    const screen = render(
      <Modal
        visible
        title="弹窗"
        content="我是弹窗"
        button={[
          { text: '取消', style: 'cancel' },
          { text: '确定' }
        ]}
      />
    )
    await waitFor(() => screen.getByText('取消'))
    await waitFor(() => screen.getByText('确定'))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('button multi', async () => {
    const screen = render(
      <Modal
        visible
        title="弹窗"
        content="我是弹窗"
        button={[
          { text: '取消', style: 'cancel' },
          { text: '自定义样式' },
          { text: '确定' }
        ]}
      />
    )
    await waitFor(() => expect(screen.container.querySelector('.xrk-modal-floor-multi')).not.toBe(null))
    await waitFor(() => screen.getByText('取消'))
    await waitFor(() => screen.getByText('确定'))
    await waitFor(() => screen.getByText('自定义样式'))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('onButtonClick', async () => {
    const onButtonClick = jest.fn()
    const screen = render(
      <Modal
        visible
        title="弹窗"
        content="我是弹窗"
        button={[
          { text: '取消', style: 'cancel' },
          { text: '确定' }
        ]}
        onButtonClick={onButtonClick}
      />
    )
    const cancelButton = screen.getByText('取消')
    const okButton = screen.getByText('确定')
    fireEvent.click(cancelButton)
    fireEvent.click(okButton)
    await waitFor(() => expect(onButtonClick.mock.calls[0][0]).toBe(0))
    await waitFor(() => expect(onButtonClick.mock.calls[1][0]).toBe(1))
    await waitFor(() => expect(onButtonClick).toHaveBeenCalledTimes(2))
  })

  test('onButtonClick not', async () => {
    const onButtonClick = jest.fn()
    const onButtonClick2 = jest.fn()
    const screen = render(
      <Modal
        visible
        title="弹窗"
        content="我是弹窗"
        button={[
          { text: '取消', style: 'cancel', onClick: onButtonClick },
          { text: '确定', onClick: onButtonClick2 }
        ]}
      />
    )
    const cancelButton = screen.getByText('取消')
    const okButton = screen.getByText('确定')
    fireEvent.click(cancelButton)
    fireEvent.click(okButton)
    await waitFor(() => expect(onButtonClick).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(onButtonClick2).toHaveBeenCalledTimes(1))
  })
})
