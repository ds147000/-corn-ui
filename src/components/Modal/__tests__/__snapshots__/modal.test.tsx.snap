// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal basis 1`] = `
<div>
  <div
    class="xrk-drawer-mask show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body show xrk-drawer-center"
    data-testid="body"
  >
    <div
      class="xrk-modal"
      data-testid="modal"
    >
      <div
        class="xrk-modal-top"
      >
        <div
          class="xrk-modal-title"
        >
          弹窗
        </div>
      </div>
      <div
        class="xrk-modal-floor-line"
      >
        <div
          class="xrk-modal-button cancel"
        >
          取消
        </div>
        <div
          class="xrk-modal-button undefined"
        >
          确定
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal button 1`] = `
<div>
  <div
    class="xrk-drawer-mask show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body show xrk-drawer-center"
    data-testid="body"
  >
    <div
      class="xrk-modal"
      data-testid="modal"
    >
      <div
        class="xrk-modal-top"
      >
        <div
          class="xrk-modal-title"
        >
          弹窗
        </div>
        <div
          class="xrk-modal-content"
        >
          我是弹窗
        </div>
      </div>
      <div
        class="xrk-modal-floor-line"
      >
        <div
          class="xrk-modal-button cancel"
        >
          取消
        </div>
        <div
          class="xrk-modal-button undefined"
        >
          确定
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal button multi 1`] = `
<div>
  <div
    class="xrk-drawer-mask show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body show xrk-drawer-center"
    data-testid="body"
  >
    <div
      class="xrk-modal"
      data-testid="modal"
    >
      <div
        class="xrk-modal-top"
      >
        <div
          class="xrk-modal-title"
        >
          弹窗
        </div>
        <div
          class="xrk-modal-content"
        >
          我是弹窗
        </div>
      </div>
      <div
        class="xrk-modal-floor-multi"
      >
        <div
          class="xrk-modal-button cancel"
        >
          取消
        </div>
        <div
          class="xrk-modal-button undefined"
        >
          自定义样式
        </div>
        <div
          class="xrk-modal-button undefined"
        >
          确定
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal children 1`] = `
<div>
  <div
    class="xrk-drawer-mask show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body show xrk-drawer-center"
    data-testid="body"
  >
    <div
      class="xrk-modal"
      data-testid="modal"
    >
      <div
        class="xrk-modal-top"
      >
        <div
          class="xrk-modal-title"
        >
          弹窗
        </div>
        <div
          class="xrk-modal-content"
        >
          <div
            class=""
          >
            我是弹窗的爸爸
          </div>
        </div>
      </div>
      <div
        class="xrk-modal-floor-line"
      >
        <div
          class="xrk-modal-button cancel"
        >
          取消
        </div>
        <div
          class="xrk-modal-button undefined"
        >
          确定
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal content 1`] = `
<div>
  <div
    class="xrk-drawer-mask show"
    data-testid="mask"
  />
  <div
    class="xrk-drawer-body show xrk-drawer-center"
    data-testid="body"
  >
    <div
      class="xrk-modal"
      data-testid="modal"
    >
      <div
        class="xrk-modal-top"
      >
        <div
          class="xrk-modal-title"
        >
          弹窗
        </div>
        <div
          class="xrk-modal-content"
        >
          我是弹窗
        </div>
      </div>
      <div
        class="xrk-modal-floor-line"
      >
        <div
          class="xrk-modal-button cancel"
        >
          取消
        </div>
        <div
          class="xrk-modal-button undefined"
        >
          确定
        </div>
      </div>
    </div>
  </div>
</div>
`;
